src/IECoreDelight/Renderer.cpp

```diff
@@ -1183,7 +1302,18 @@ class DelightRenderer final : public IECoreScenePreview::Renderer
 
 		ObjectInterfacePtr light( const std::string &name, const IECore::Object *object, const AttributesInterface *attributes ) override
 		{
-			return this->object( name, object, attributes );
+			const IECore::MessageHandler::Scope s( m_messageHandler.get() );
+
+			DelightHandleSharedPtr instance;
+			if( object )
+			{
+				instance = m_instanceCache->get( object );
+			}
+
+			ObjectInterfacePtr result = new DelightLight( m_context, name, instance, ownership() );
```

Comment:
> In the previous version I think you fell back to creating a DelightObject in this case?

Yes, that's correct. My goal is to continue to support existing 3Delight native lights, for backwards compatibility and since they have additional parameters USD lights don't yet support. For those we're getting an OSLLight which does the geometry creation for us and passes it via `object`, which I'm converting to 3Delight geometry and using that as the `DelightLight::m_instance`. That's pretty much the same as how the previous lights were working.

In that case, it is effectively falling back to `DelightObject` because all of the special handling in `DelightLight::attributes()` is behind a check to see if it's a USD light.

I don't think I'm following when we'd want to have both implicit and explicit geometry. Implicit for a USD RectLight, for example, but when would Gaffer try and send geometry to a RectLight?

Reply:
In https://github.com/GafferHQ/gaffer/pull/5603/commits/a0c61d62e4bb80bef1f2ddfe3839f68389fcb36d, I added the additional `DelightHandleSharedPtr` for the light geometry as you described for maximum flexibility.